Plugin isn't compatible with Custom Tools

This is similar to #19... it seems that this plugin isn't compatible with the Custom Tools plugin.

I have the Custom Tools plugin configured as follows:

![tools](https://cloud.githubusercontent.com/assets/234676/7999884/60195c60-0b96-11e5-9752-e87748b0bd95.png)

The output from the job is as follows:

```
00:00:00.001 [EnvInject] - Inject global passwords.
00:00:00.002 Started by remote host 10.182.88.66
00:00:00.002 [CustomTools] - arcanist: Starting installation
00:00:00.009 [arcanist] $ sh -e /mnt/jenkins-slave-workspace/tools/com.cloudbees.jenkins.plugins.customtools.CustomTool/arcanist/hudson8373437206605157068.sh
00:00:00.023 [CustomTools] - arcanist: Tool is installed at /mnt/jenkins-slave-workspace/tools/com.cloudbees.jenkins.plugins.customtools.CustomTool/arcanist
00:00:00.023 [CustomTools] - arcanist: Setting arcanist_HOME=/mnt/jenkins-slave-workspace/tools/com.cloudbees.jenkins.plugins.customtools.CustomTool/arcanist
00:00:00.023 [EnvInject] - Loading node environment variables.
...
00:00:00.623 [redacted] $ sh -c 'echo '{"ids":["558494"]}' | arc call-conduit differential.querydiffs'
00:00:01.033 Applying patch for differential
00:00:01.036 [redacted] $ sh -c 'echo '{"silent":true,"action":"none","message":"Build started: https://redacted/job/redacted/19/ (console: https://redacted/job/redacted/19/console)","revision_id":"20969"}' | arc call-conduit differential.createcomment'
00:00:01.569 [redacted] $ git reset --hard origin/master
00:00:01.579 HEAD is now at 02a3a6c chore(build): bump version
00:00:01.585 [redacted] $ git clean -fd -f
00:00:01.597 [redacted] $ git submodule update --init --recursive
00:00:01.661 [redacted] $ arc patch --nobranch --diff 558494 --nocommit
00:00:01.663 FATAL: Cannot run program "arc" (in directory "/mnt/jenkins-slave-workspace/workspace/redacted"): error=2, No such file or directory
00:00:01.664 java.io.IOException: Cannot run program "arc" (in directory "/mnt/jenkins-slave-workspace/workspace/redacted"): error=2, No such file or directory
00:00:01.664    at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
00:00:01.664    at hudson.Proc$LocalProc.<init>(Proc.java:244)
00:00:01.664    at hudson.Proc$LocalProc.<init>(Proc.java:216)
00:00:01.664    at hudson.Launcher$LocalLauncher.launch(Launcher.java:816)
00:00:01.664    at hudson.Launcher$ProcStarter.start(Launcher.java:382)
00:00:01.664    at hudson.Launcher$RemoteLaunchCallable.call(Launcher.java:1149)
00:00:01.664    at hudson.Launcher$RemoteLaunchCallable.call(Launcher.java:1114)
00:00:01.664    at hudson.remoting.UserRequest.perform(UserRequest.java:121)
00:00:01.664    at hudson.remoting.UserRequest.perform(UserRequest.java:49)
00:00:01.664    at hudson.remoting.Request$2.run(Request.java:324)
00:00:01.664    at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
00:00:01.664    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
00:00:01.664    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
00:00:01.664    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
00:00:01.664    at hudson.remoting.Engine$1$1.run(Engine.java:63)
00:00:01.664    at java.lang.Thread.run(Thread.java:745)
00:00:01.664    at ......remote call to i-b7ea994a-ef4326cf(Native Method)
00:00:01.664    at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1361)
00:00:01.664    at hudson.remoting.UserResponse.retrieve(UserRequest.java:221)
00:00:01.664    at hudson.remoting.Channel.call(Channel.java:753)
00:00:01.664    at hudson.Launcher$RemoteLauncher.launch(Launcher.java:929)
00:00:01.664    at com.cloudbees.jenkins.plugins.customtools.CustomToolInstallWrapper$2.launch(CustomToolInstallWrapper.java:255)
00:00:01.664    at hudson.Launcher$ProcStarter.start(Launcher.java:382)
00:00:01.664    at hudson.Launcher$ProcStarter.join(Launcher.java:389)
00:00:01.664    at com.uber.jenkins.phabricator.PhabricatorBuildWrapper.setUp(PhabricatorBuildWrapper.java:120)
00:00:01.664    at hudson.model.Build$BuildExecution.doRun(Build.java:156)
00:00:01.664    at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:537)
00:00:01.664    at hudson.model.Run.execute(Run.java:1744)
00:00:01.664    at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
00:00:01.664    at hudson.model.ResourceController.execute(ResourceController.java:98)
00:00:01.664    at hudson.model.Executor.run(Executor.java:374)
00:00:01.665 Caused by: java.io.IOException: error=2, No such file or directory
00:00:01.665    at java.lang.UNIXProcess.forkAndExec(Native Method)
00:00:01.665    at java.lang.UNIXProcess.<init>(UNIXProcess.java:248)
00:00:01.665    at java.lang.ProcessImpl.start(ProcessImpl.java:134)
00:00:01.665    at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
00:00:01.665    at hudson.Proc$LocalProc.<init>(Proc.java:244)
00:00:01.665    at hudson.Proc$LocalProc.<init>(Proc.java:216)
00:00:01.665    at hudson.Launcher$LocalLauncher.launch(Launcher.java:816)
00:00:01.665    at hudson.Launcher$ProcStarter.start(Launcher.java:382)
00:00:01.665    at hudson.Launcher$RemoteLaunchCallable.call(Launcher.java:1149)
00:00:01.665    at hudson.Launcher$RemoteLaunchCallable.call(Launcher.java:1114)
00:00:01.665    at hudson.remoting.UserRequest.perform(UserRequest.java:121)
00:00:01.665    at hudson.remoting.UserRequest.perform(UserRequest.java:49)
00:00:01.665    at hudson.remoting.Request$2.run(Request.java:324)
00:00:01.665    at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
00:00:01.665    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
00:00:01.665    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
00:00:01.665    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
00:00:01.665    at hudson.remoting.Engine$1$1.run(Engine.java:63)
00:00:01.665    at java.lang.Thread.run(Thread.java:745)
```

--- a/src/main/java/com/uber/jenkins/phabricator/PhabricatorBuildWrapper.java
+++ b/src/main/java/com/uber/jenkins/phabricator/PhabricatorBuildWrapper.java
@@ -54,8 +54,11 @@ public class PhabricatorBuildWrapper extends BuildWrapper {
             return this.ignoreBuild(logger, "No environment variables found?!");
         }
 
+        final String arcPath = this.getArcPath();
+
         final Map<String, String> envAdditions = new HashMap<String, String>();
         envAdditions.put(PhabricatorPlugin.WRAP_KEY, "true");
+        envAdditions.put(PhabricatorPlugin.ARCANIST_PATH, arcPath);
 
         final String conduitToken = this.getConduitToken();
         if (!CommonUtils.isBlank(conduitToken)) {
@@ -81,7 +84,7 @@ public class PhabricatorBuildWrapper extends BuildWrapper {
                 }
             }
 
-            Differential diff = Differential.fromDiffID(diffID, starter, conduitToken);
+            Differential diff = Differential.fromDiffID(diffID, starter, conduitToken, arcPath);
             diff.decorate(build, this.getPhabricatorURL());
 
             logger.println("Applying patch for differential");
@@ -115,7 +118,7 @@ public class PhabricatorBuildWrapper extends BuildWrapper {
                     .cmds(Arrays.asList("git", "submodule", "update", "--init", "--recursive"))
                     .join();
 
-            List<String> patchCommand = new ArrayList<String>(Arrays.asList("arc", "patch", "--nobranch", "--diff", diffID));
+            List<String> patchCommand = new ArrayList<String>(Arrays.asList(arcPath, "patch", "--nobranch", "--diff", diffID));
             if (!createCommit) {
                 patchCommand.add("--nocommit");
             }
@@ -180,6 +183,10 @@ public class PhabricatorBuildWrapper extends BuildWrapper {
 
     public String getConduitToken() {
         return this.getDescriptor().getConduitToken();
+    }
+
+    public String getArcPath() {
+        return this.getDescriptor().getArcPath();
     }
 
     // Overridden for better type safety.

--- a/src/main/java/com/uber/jenkins/phabricator/PhabricatorBuildWrapper.java
+++ b/src/main/java/com/uber/jenkins/phabricator/PhabricatorBuildWrapper.java
@@ -84,7 +84,7 @@ public class PhabricatorBuildWrapper extends BuildWrapper {
                 }
             }
 
-            Differential diff = Differential.fromDiffID(diffID, starter, conduitToken, arcPath);
+            Differential diff = new Differential(diffID, starter, conduitToken, arcPath);
             diff.decorate(build, this.getPhabricatorURL());
 
             logger.println("Applying patch for differential");

--- a/src/main/java/com/uber/jenkins/phabricator/PhabricatorBuildWrapper.java
+++ b/src/main/java/com/uber/jenkins/phabricator/PhabricatorBuildWrapper.java
@@ -185,8 +185,16 @@ public class PhabricatorBuildWrapper extends BuildWrapper {
         return this.getDescriptor().getConduitToken();
     }
 
+    /**
+     * Return the path to the arcanist executable
+     * @return a string, fully-qualified or not, could just be "arc"
+     */
     public String getArcPath() {
-        return this.getDescriptor().getArcPath();
+        final String providedPath = this.getDescriptor().getArcPath();
+        if (CommonUtils.isBlank(providedPath)) {
+            return "arc";
+        }
+        return providedPath;
     }
 
     // Overridden for better type safety.

--- a/src/main/java/com/uber/jenkins/phabricator/PhabricatorBuildWrapperDescriptor.java
+++ b/src/main/java/com/uber/jenkins/phabricator/PhabricatorBuildWrapperDescriptor.java
@@ -32,6 +32,7 @@ import org.kohsuke.stapler.StaplerRequest;
 public final class PhabricatorBuildWrapperDescriptor extends BuildWrapperDescriptor {
     private String conduitURL;
     private String conduitToken;
+    private String arcPath;
 
     public PhabricatorBuildWrapperDescriptor() {
         super(PhabricatorBuildWrapper.class);
@@ -69,5 +70,12 @@ public final class PhabricatorBuildWrapperDescriptor extends BuildWrapperDescrip
     }
     public void setConduitToken(String conduitToken) {
         this.conduitToken = conduitToken;
+    }
+
+    public String getArcPath() {
+        return arcPath;
+    }
+    public void setArcPath(String arcPath) {
+        this.arcPath = arcPath;
     }
 }

--- a/src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java
+++ b/src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java
@@ -114,7 +114,7 @@ public class PhabricatorNotifier extends Notifier {
 
         LauncherFactory starter = new LauncherFactory(launcher, environment, listener.getLogger(), build.getWorkspace());
 
-        Differential diff = Differential.fromDiffID(diffID, starter, conduitToken, arcPath);
+        Differential diff = new Differential(diffID, starter, conduitToken, arcPath);
 
         String revisionID = diff.getRevisionID();
         if (CommonUtils.isBlank(revisionID)) {

--- a/src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java
+++ b/src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java
@@ -82,8 +82,9 @@ public class PhabricatorNotifier extends Notifier {
         }
 
         UberallsClient uberalls = new UberallsClient(getDescriptor().getUberallsURL(), environment, logger);
-        boolean needsDecoration = environment.get(PhabricatorPlugin.WRAP_KEY, null) == null;
-        String conduitToken = environment.get(PhabricatorPlugin.CONDUIT_TOKEN, null);
+        final boolean needsDecoration = environment.get(PhabricatorPlugin.WRAP_KEY, null) == null;
+        final String conduitToken = environment.get(PhabricatorPlugin.CONDUIT_TOKEN, null);
+        final String arcPath = environment.get(PhabricatorPlugin.ARCANIST_PATH, "arc");
 
         boolean uberallsConfigured = !CommonUtils.isBlank(uberalls.getBaseURL());
 
@@ -113,7 +114,7 @@ public class PhabricatorNotifier extends Notifier {
 
         LauncherFactory starter = new LauncherFactory(launcher, environment, listener.getLogger(), build.getWorkspace());
 
-        Differential diff = Differential.fromDiffID(diffID, starter, conduitToken);
+        Differential diff = Differential.fromDiffID(diffID, starter, conduitToken, arcPath);
 
         String revisionID = diff.getRevisionID();
         if (CommonUtils.isBlank(revisionID)) {

--- a/src/main/java/com/uber/jenkins/phabricator/PhabricatorPlugin.java
+++ b/src/main/java/com/uber/jenkins/phabricator/PhabricatorPlugin.java
@@ -26,13 +26,15 @@ import jenkins.model.Jenkins;
 
 import java.io.File;
 
-public class PhabricatorPlugin extends Plugin {    // Diff ID (not differential ID)
+public class PhabricatorPlugin extends Plugin {
+    // Diff ID (not differential ID)
     static final String DIFFERENTIAL_ID_FIELD = "DIFF_ID";
     // Phabricator object ID (for Harbormaster)
     static final String PHID_FIELD = "PHID";
 
     static final String WRAP_KEY = "PHABRICATOR_JENKINS";
     public static final String CONDUIT_TOKEN = "CONDUIT_TOKEN";
+    public static final String ARCANIST_PATH = "ARCANIST_PATH";
 
     public static String getIconPath(String icon) {
         if(icon == null) return null;

--- a/src/main/java/com/uber/jenkins/phabricator/conduit/ArcanistClient.java
+++ b/src/main/java/com/uber/jenkins/phabricator/conduit/ArcanistClient.java
@@ -28,24 +28,26 @@ import net.sf.json.groovy.JsonSlurper;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.PrintStream;
-import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 
 public class ArcanistClient {
+    private final String arcPath;
     private final String methodName;
     private final Map<String, String> params;
     private final String conduitToken;
 
-    public ArcanistClient(String methodName, Map<String, String> params, String conduitToken) {
+    public ArcanistClient(String arcPath, String methodName, Map<String, String> params, String conduitToken) {
+        this.arcPath = arcPath;
         this.methodName = methodName;
         this.params = params;
         this.conduitToken = conduitToken;
     }
 
     private String getConduitCommand(String methodName) {
-        StringBuilder sb = new StringBuilder("arc call-conduit ");
+        StringBuilder sb = new StringBuilder(this.arcPath);
+        sb.append(" call-conduit ");
         sb.append(methodName);
         if (!CommonUtils.isBlank(this.conduitToken)) {
             sb.append(" --conduit-token=");

--- a/src/main/java/com/uber/jenkins/phabricator/conduit/Differential.java
+++ b/src/main/java/com/uber/jenkins/phabricator/conduit/Differential.java
@@ -46,8 +46,8 @@ public class Differential {
         Map params = new HashMap<String, String>();
         params.put("ids", new String[]{diffID});
 
-        ArcanistClient arc = this.createClient( "differential.querydiffs", params);
-        this.rawJSON = (JSONObject) ((JSONObject) arc.callConduit(launcher.launch(), launcher.getStderr()).get("response")).get(diffID);
+        JSONObject query = this.callConduit("differential.querydiffs", params);
+        this.rawJSON = (JSONObject) ((JSONObject) query.get("response")).get(diffID);
     }
 
     public String getRevisionID(boolean formatted) {
@@ -77,8 +77,12 @@ public class Differential {
         params.put("type", pass ? "pass" : "fail");
         params.put("buildTargetPHID", phid);
 
-        ArcanistClient arc = this.createClient("harbormaster.sendmessage", params);
-        arc.callConduit(this.launcher.launch(), this.launcher.getStderr());
+        this.callConduit("harbormaster.sendmessage", params);
+    }
+
+    private JSONObject callConduit(String methodName, Map<String, String> params) throws IOException, InterruptedException {
+        ArcanistClient arc = new ArcanistClient(this.arcPath, methodName, params, this.conduitToken);
+        return arc.callConduit(this.launcher.launch(), this.launcher.getStderr());
     }
 
     /**
@@ -94,12 +98,7 @@ public class Differential {
         params.put("message", this.escapeSpecialCharacters(message));
         params.put("silent", silent);
 
-        ArcanistClient arc = this.createClient("differential.createcomment", params);
-        return arc.callConduit(this.launcher.launch(), this.launcher.getStderr());
-    }
-
-    private ArcanistClient createClient(String methodName, Map<String, String> params) {
-        return new ArcanistClient(this.arcPath, methodName, params, this.conduitToken);
+        return this.callConduit("differential.createcomment", params);
     }
 
     public JSONObject postComment(String message) throws IOException, InterruptedException {

--- a/src/main/java/com/uber/jenkins/phabricator/conduit/Differential.java
+++ b/src/main/java/com/uber/jenkins/phabricator/conduit/Differential.java
@@ -37,6 +37,7 @@ public class Differential {
     private final JSONObject rawJSON;
     private final LauncherFactory launcher;
     private final String conduitToken;
+    private final String arcPath;
 
     /**
      * Instantiate a differential from a diff ID (not differential ID)
@@ -44,24 +45,26 @@ public class Differential {
      * @param launcher
      * @return
      */
-    public static Differential fromDiffID(String diffID, LauncherFactory launcher, String conduitToken) throws IOException, InterruptedException {
+    public static Differential fromDiffID(String diffID, LauncherFactory launcher, String conduitToken, String arcPath) throws IOException, InterruptedException {
         Map params = new HashMap<String, String>();
         params.put("ids", new String[]{diffID});
-        ArcanistClient arc = new ArcanistClient("differential.querydiffs", params, conduitToken);
+        ArcanistClient arc = new ArcanistClient(arcPath, "differential.querydiffs", params, conduitToken);
 
         Differential diff = new Differential(
                 (JSONObject) ((JSONObject) arc.callConduit(launcher.launch(), launcher.getStderr()).get("response")).get(diffID),
                 launcher,
-                conduitToken
+                conduitToken,
+                arcPath
         );
 
         return diff;
     }
 
-    Differential(JSONObject rawJSON, LauncherFactory launcher, String conduitToken) {
+    Differential(JSONObject rawJSON, LauncherFactory launcher, String conduitToken, String arcPath) {
         this.rawJSON = rawJSON;
         this.launcher = launcher;
         this.conduitToken = conduitToken;
+        this.arcPath = arcPath;
     }
 
     public String getRevisionID(boolean formatted) {
@@ -90,7 +93,7 @@ public class Differential {
         Map params = new HashMap<String, String>();
         params.put("type", pass ? "pass" : "fail");
         params.put("buildTargetPHID", phid);
-        ArcanistClient arc = new ArcanistClient("harbormaster.sendmessage", params, this.conduitToken);
+        ArcanistClient arc = new ArcanistClient(this.arcPath, "harbormaster.sendmessage", params, this.conduitToken);
 
         arc.callConduit(this.launcher.launch(), this.launcher.getStderr());
     }
@@ -108,9 +111,12 @@ public class Differential {
         params.put("message", this.escapeSpecialCharacters(message));
         params.put("silent", silent);
 
-        ArcanistClient arc = new ArcanistClient("differential.createcomment", params, this.conduitToken);
-
+        ArcanistClient arc = this.createClient("differential.createcomment", params);
         return arc.callConduit(this.launcher.launch(), this.launcher.getStderr());
+    }
+
+    private ArcanistClient createClient(String methodName, Map<String, String> params) {
+        return new ArcanistClient(this.arcPath, methodName, params, this.conduitToken);
     }
 
     public JSONObject postComment(String message) throws IOException, InterruptedException {

--- a/src/main/java/com/uber/jenkins/phabricator/conduit/Differential.java
+++ b/src/main/java/com/uber/jenkins/phabricator/conduit/Differential.java
@@ -39,32 +39,15 @@ public class Differential {
     private final String conduitToken;
     private final String arcPath;
 
-    /**
-     * Instantiate a differential from a diff ID (not differential ID)
-     * @param diffID
-     * @param launcher
-     * @return
-     */
-    public static Differential fromDiffID(String diffID, LauncherFactory launcher, String conduitToken, String arcPath) throws IOException, InterruptedException {
-        Map params = new HashMap<String, String>();
-        params.put("ids", new String[]{diffID});
-        ArcanistClient arc = new ArcanistClient(arcPath, "differential.querydiffs", params, conduitToken);
-
-        Differential diff = new Differential(
-                (JSONObject) ((JSONObject) arc.callConduit(launcher.launch(), launcher.getStderr()).get("response")).get(diffID),
-                launcher,
-                conduitToken,
-                arcPath
-        );
-
-        return diff;
-    }
-
-    Differential(JSONObject rawJSON, LauncherFactory launcher, String conduitToken, String arcPath) {
-        this.rawJSON = rawJSON;
-        this.launcher = launcher;
+    public Differential(String diffID, LauncherFactory launcher, String conduitToken, String arcPath) throws IOException, InterruptedException {
         this.conduitToken = conduitToken;
         this.arcPath = arcPath;
+        this.launcher = launcher;
+        Map params = new HashMap<String, String>();
+        params.put("ids", new String[]{diffID});
+
+        ArcanistClient arc = this.createClient( "differential.querydiffs", params);
+        this.rawJSON = (JSONObject) ((JSONObject) arc.callConduit(launcher.launch(), launcher.getStderr()).get("response")).get(diffID);
     }
 
     public String getRevisionID(boolean formatted) {
@@ -93,8 +76,8 @@ public class Differential {
         Map params = new HashMap<String, String>();
         params.put("type", pass ? "pass" : "fail");
         params.put("buildTargetPHID", phid);
-        ArcanistClient arc = new ArcanistClient(this.arcPath, "harbormaster.sendmessage", params, this.conduitToken);
 
+        ArcanistClient arc = this.createClient("harbormaster.sendmessage", params);
         arc.callConduit(this.launcher.launch(), this.launcher.getStderr());
     }
 

--- a/src/main/resources/com/uber/jenkins/phabricator/PhabricatorBuildWrapper/global.jelly
+++ b/src/main/resources/com/uber/jenkins/phabricator/PhabricatorBuildWrapper/global.jelly
@@ -20,5 +20,9 @@
       description="Conduit token for Phabricator API. If left blank, reads from ~/.arcrc">
       <f:textbox />
     </f:entry>
+    <f:entry title="Location of arcanist" field="arcPath"
+      description="location of `arc` binary, if not in path">
+      <f:textbox default="arc" />
+    </f:entry>
   </f:section>
 </j:jelly>

